### DISCUSSION TITLE
cross-plugin dependency checking

### DISCUSSION THREAD

**Discussion by jrhender** [decentralized-identity/veramo#699] [Q&A] [answered]
Is there currently a way to do validation of cross-plugin dependencies at compile time (typescript checking) or a build time (e.g. running a tool against your agent configuration)?

For instance, if there is an agent factory like this (with the `KeyManager` commented to illustrate the issue and with `imports` and other config omitted for brevity):
```ts
export const createAgentWithProvider = (
  provider: providers.JsonRpcProvider
) => {
  return createAgent<IDIDManager & IKeyManager & IResolver & IDataStore>({
    plugins: [
      // new KeyManager({
      //   store: new KeyStore(dbConnection),
      //   kms: {
      //     local: new KeyManagementSystem(),
      //   },
      // }),
      new DIDManager({
        store: new DIDStore(dbConnection),
        defaultProvider: "did:ethr:volta",
        providers: {
          "did:ethr:volta": new EthrDIDProvider({
            defaultKms: "local",
            network: "volta",
            web3Provider: provider,
          }),
        },
      })
    ],
  });
};
```
And construct the agent and execute a method like this
```ts
const provider = new providers.JsonRpcProvider(
    "https://volta-rpc.energyweb.org"
  );
const agent = createAgentWithProvider(provider);
const identifier = await agent.didManagerCreate();
console.log(identifier);
```
then at runtime I get this error:
```
(node:72138) UnhandledPromiseRejectionWarning: TypeError: context.agent.keyManagerCreate is not a function
    at EthrDIDProvider.<anonymous> (/Users/john/ew/veramo-hackathon/veramo-agent/node_modules/@veramo/did-provider-ethr/src/ethr-did-provider.ts:61:37)
```
Is there currently a way to check ahead of runtime that `keyManagerCreate` is available on the agent's methods (because it is needed as a dependency of `EthrDIDProvider`)?

**Comment by mirceanis** [comment, marked as answer]
That's an excellent question!
This is actually a feature that we planned from the beginning, but never got enough time for implementation.
So, at this point in time, the answer is no.
It might be difficult to have build time checks, at least initially, but early runtime checks should be possible with minimal changes.

At the moment, plugins can declare a schema that they provide, but they don't provide a set of requirements. I think these requirements need to be declared for the checks to be made.

Feel free to post ideas here if you've got a clever way to run these checks other that what I'm suggesting.

**Reply by jrhender** [reply]
Thanks for this answer. I don't have an especially clever idea to do it at the moment. I was thinking that perhaps the schema declaration files could be used as you suggested but haven't thought about it much yet. But it's good to know what the current possibilities are 👍
